File: core/src/test/java/dev/morphia/critter/parser/gizmo/TestGizmoGeneration.java
package dev.morphia.critter.parser.gizmo;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import dev.morphia.annotations.Entity;
import dev.morphia.annotations.EntityListeners;
import dev.morphia.annotations.Indexes;
import dev.morphia.annotations.internal.CollationBuilder;
import dev.morphia.annotations.internal.EntityBuilder;
import dev.morphia.annotations.internal.EntityListenersBuilder;
import dev.morphia.annotations.internal.FieldBuilder;
import dev.morphia.annotations.internal.IndexBuilder;
import dev.morphia.annotations.internal.IndexOptionsBuilder;
import dev.morphia.annotations.internal.IndexesBuilder;
import dev.morphia.config.MorphiaConfig;
import dev.morphia.critter.ClassfileOutput;
import dev.morphia.critter.CritterClassLoader;
import dev.morphia.critter.parser.Generators;
import dev.morphia.critter.parser.asm.AddMethodAccessorMethods;
import dev.morphia.critter.sources.Example;
import dev.morphia.critter.sources.MethodExample;
import dev.morphia.mapping.ReflectiveMapper;
import dev.morphia.mapping.codec.pojo.EntityModel;
import dev.morphia.mapping.codec.pojo.PropertyModel;
import dev.morphia.mapping.codec.pojo.TypeData;
import dev.morphia.mapping.lifecycle.EntityListenerAdapter;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.AnnotationNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;
import org.testng.Assert;
import org.testng.annotations.Test;

import io.quarkus.gizmo.ClassCreator;
import io.quarkus.gizmo.MethodDescriptor;

import static com.mongodb.client.model.CollationCaseFirst.LOWER;
import static io.quarkus.gizmo.MethodDescriptor.ofMethod;

public class TestGizmoGeneration {
    private final CritterClassLoader critterClassLoader = new CritterClassLoader();

    @Test
    public void testMapStringExample() {
        String descString = "Ljava/util/Map<Ljava/lang/String;Ldev/morphia/critter/sources/Example;>;";
        String descriptor = descriptor(
                java.util.Map.class,
                descriptor(String.class),
                descriptor(Example.class));

        Assert.assertEquals(descriptor, descString);
        TypeData<?> typeData = PropertyModelGenerator.typeData(descString, Thread.currentThread().getContextClassLoader()).get(0);
        Assert.assertEquals(
                typeData,
                typeDataHelper(java.util.Map.class, typeDataHelper(String.class), typeDataHelper(Example.class)));
    }

    @Test
    public void testListMapStringExample() {
        String descString = "Ljava/util/List<Ljava/util/Map<Ljava/lang/String;Ldev/morphia/critter/sources/Example;>;>;";
        String descriptor = descriptor(
                java.util.List.class,
                descriptor(
                        java.util.Map.class,
                        descriptor(String.class),
                        descriptor(Example.class)));
        Assert.assertEquals(descriptor, descString);

        TypeData<?> typeData = PropertyModelGenerator.typeData(descString, Thread.currentThread().getContextClassLoader()).get(0);
        Assert.assertEquals(
                typeData,
                typeDataHelper(java.util.List.class,
                        typeDataHelper(java.util.Map.class, typeDataHelper(String.class), typeDataHelper(Example.class))));
    }

    @Test
    public void testMapOfList() {
        String descString = "Ljava/util/Map<Ljava/lang/String;Ljava/util/List<Ldev/morphia/critter/sources/Example;>;>;";
        String descriptor = descriptor(
                java.util.Map.class,
                descriptor(String.class),
                descriptor(java.util.List.class, descriptor(Example.class)));
        Assert.assertEquals(descriptor, descString);
        TypeData<?> typeData = PropertyModelGenerator.typeData(descriptor, Thread.currentThread().getContextClassLoader()).get(0);
        Assert.assertEquals(
                typeData,
                typeDataHelper(java.util.Map.class,
                        typeDataHelper(String.class),
                        typeDataHelper(java.util.List.class, typeDataHelper(Example.class))));
    }

    @Test
    public void testPrimitiveArray() {
        TypeData<?> typeData = PropertyModelGenerator.typeData("[I", Thread.currentThread().getContextClassLoader()).get(0);
        Assert.assertTrue(typeData.isArray());
    }

    @Test
    public void testAnnotationBuilding() throws Exception {
        AnnotationNode index = new AnnotationNode("Ldev/morphia/annotations/Index;");
        AnnotationNode field = new AnnotationNode("Ldev/morphia/annotations/Field;");
        index.values = List.of("fields", List.of(field));

        try (ClassCreator creator = ClassCreator.builder()
                .className("critter.AnnotationTest")
                .superClass(EntityModel.class)
                .classOutput((name, data) -> {
                    String className = name.replace('/', '.');
                    critterClassLoader.register(className, data);
                    try {
                        ClassfileOutput.dump(name, data);
                    } catch (Exception ignored) {
                    }
                })
                .build()) {
            var mc = creator.getMethodCreator("test", Void.class);
            MethodDescriptor annotationMethod = ofMethod(
                    EntityModel.class.getName(),
                    "annotation",
                    EntityModel.class.getName(),
                    java.lang.annotation.Annotation.class);
            mc.invokeVirtualMethod(
                    annotationMethod,
                    mc.getThis(),
                    GizmoExtensions.annotationBuilder(index, mc));
        }
    }

    @Test
    public void testGizmo() throws Exception {
        MorphiaConfig config = dev.morphia.config.MorphiaConfig.load();
        Generators generators = new Generators(config, new ReflectiveMapper(config));
        CritterGizmoGenerator.generate(Example.class, critterClassLoader, generators, false);
        critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.AgeModel");
        Class<?> nameModel = critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.NameModel");
        invokeAll(PropertyModel.class, nameModel);
        critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.SalaryModel");
        critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.AgeAccessor").getConstructor().newInstance();
        critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.NameAccessor").getConstructor().newInstance();
        critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.SalaryAccessor").getConstructor().newInstance();

        Class<?> loadClass = critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.ExampleEntityModel");
        EntityModel model = (EntityModel) loadClass.getConstructors()[0].newInstance(generators.getMapper());
        validate(model);
    }

    private void validate(EntityModel model) {
        Assert.assertEquals(
                model.getAnnotation(EntityListeners.class),
                EntityListenersBuilder.entityListenersBuilder().value(EntityListenerAdapter.class).build());
        Assert.assertEquals(
                model.getAnnotation(Entity.class),
                EntityBuilder.entityBuilder().value("examples").build());
        Assert.assertEquals(
                model.getAnnotation(Indexes.class),
                IndexesBuilder.indexesBuilder()
                        .value(IndexBuilder.indexBuilder()
                                .fields(FieldBuilder.fieldBuilder().value("name").weight(42).build())
                                .options(IndexOptionsBuilder.indexOptionsBuilder()
                                        .partialFilter("partial filter")
                                        .collation(CollationBuilder.collationBuilder().caseFirst(LOWER).build())
                                        .build())
                                .build())
                        .build());
        Assert.assertEquals(model.collectionName(), "examples");
        Assert.assertEquals(model.discriminator(), "Example");
        Assert.assertEquals(model.discriminatorKey(), "_t");
        Assert.assertEquals(model.getType().getName(), Example.class.getName());
        Assert.assertFalse(model.getProperties().isEmpty(), "Should have properties");
        Assert.assertNotNull(model.getIdProperty(), "Should have an ID property");
        Assert.assertFalse(model.isAbstract(), "Should not be abstract");
        Assert.assertFalse(model.isInterface(), "Should not be an interface");
        Assert.assertTrue(model.useDiscriminator(), "Should use the discriminator");
        Assert.assertTrue(model.classHierarchy().isEmpty(), "Should not have a class hierarchy");
    }

    private void invokeAll(Class<?> type, Class<?> klass) {
        Object instance;
        try {
            instance = klass.getConstructors()[0].newInstance(new Object[] { null });
        } catch (Exception e) {
            Assert.fail("Could not instantiate " + klass.getName() + ": " + e.getMessage());
            return;
        }
        List<String> results = Arrays.stream(type.getDeclaredMethods())
                .filter(m -> Modifier.isPublic(m.getModifiers())
                        && !Modifier.isFinal(m.getModifiers())
                        && m.getParameterCount() == 0)
                .filter(m -> !List.of("hashCode", "toString").contains(m.getName()))
                .sorted(Comparator.comparing(Method::getName))
                .map(method -> {
                    try {
                        klass.getDeclaredMethod(method.getName(), method.getParameterTypes());
                        return null;
                    } catch (Exception e) {
                        return e.getMessage();
                    }
                })
                .filter(Objects::nonNull)
                .collect(Collectors.toList());

        if (!results.isEmpty()) {
            Assert.fail("Missing methods from " + type.getName() + ": \n" + String.join("\n", results));
        }
    }

    @Test
    public void testConstructors() throws Exception {
        String className = "dev.morphia.critter.GizmoSubclass";

        try (ClassCreator constructorCall = ClassCreator.builder()
                .classOutput((name, data) -> critterClassLoader.register(name.replace('/', '.'), data))
                .className("dev.morphia.critter.ConstructorCall")
                .build()) {
            var fieldCreator = constructorCall.getFieldCreator("name", String.class)
                    .setModifiers(Modifier.PUBLIC);
            var constructorCreator = constructorCall.getConstructorCreator(String.class);
            constructorCreator.invokeSpecialMethod(
                    MethodDescriptor.ofConstructor(Object.class),
                    constructorCreator.getThis());
            constructorCreator.setParameterNames(new String[] { "name" });
            constructorCreator.writeInstanceField(
                    fieldCreator.getFieldDescriptor(),
                    constructorCreator.getThis(),
                    constructorCreator.getMethodParam(0));
            constructorCreator.returnVoid();
        }

        critterClassLoader.loadClass("dev.morphia.critter.ConstructorCall")
                .getConstructor(String.class)
                .newInstance("here i am");

        try (ClassCreator creator = ClassCreator.builder()
                .classOutput((name, data) -> critterClassLoader.register(name.replace('/', '.'), data))
                .className(className)
                .superClass("dev.morphia.critter.ConstructorCall")
                .build()) {
            var constructor = creator.getConstructorCreator(String.class);
            constructor.invokeSpecialMethod(
                    MethodDescriptor.ofConstructor("dev.morphia.critter.ConstructorCall", String.class),
                    constructor.getThis(),
                    constructor.getMethodParam(0));
            constructor.setParameterNames(new String[] { "subName" });
            constructor.returnVoid();
        }

        Object instance = critterClassLoader.loadClass(className)
                .getConstructor(String.class)
                .newInstance("This is my name");
        Assert.assertNotNull(instance);
    }

    @Test
    public void testMethodBasedAccessors() throws Exception {
        CritterClassLoader classLoader = new CritterClassLoader();

        String resourceName = MethodExample.class.getName().replace('.', '/') + ".class";
        var inputStream = MethodExample.class.getClassLoader().getResourceAsStream(resourceName);
        ClassNode classNode = new ClassNode();
        new ClassReader(inputStream).accept(classNode, 0);

        List<MethodNode> methodNodes = classNode.methods.stream()
                .filter(node -> node.name.startsWith("get"))
                .filter(node -> Type.getArgumentTypes(node.desc).length == 0)
                .filter(node -> node.visibleAnnotations != null
                        && node.visibleAnnotations.stream().anyMatch(
                                ann -> List.of("Ldev/morphia/annotations/Id;", "Ldev/morphia/annotations/Property;").contains(ann.desc)))
                .collect(Collectors.toList());

        List<String> methodNames = methodNodes.stream().map(n -> n.name).collect(Collectors.toList());
        Assert.assertTrue(methodNames.contains("getId"), "Should find getId method");
        Assert.assertTrue(methodNames.contains("getCount"), "Should find getCount method");
        Assert.assertTrue(methodNames.contains("getScore"), "Should find getScore method");
        Assert.assertTrue(methodNames.contains("getComputedValue"), "Should find getComputedValue method");
        Assert.assertEquals(4, methodNodes.size(), "Should find exactly 4 annotated getter methods");

        byte[] bytecode = new AddMethodAccessorMethods(MethodExample.class, methodNodes).emit();

        classLoader.register(MethodExample.class.getName(), bytecode);
        Class<?> modifiedClass = classLoader.loadClass(MethodExample.class.getName());

        Assert.assertNotNull(modifiedClass.getMethod("__readId"), "Should have __readId method");
        Assert.assertNotNull(modifiedClass.getMethod("__readCount"), "Should have __readCount method");
        Assert.assertNotNull(modifiedClass.getMethod("__readScore"), "Should have __readScore method");
        Assert.assertNotNull(modifiedClass.getMethod("__readComputedValue"), "Should have __readComputedValue method");

        Assert.assertNotNull(modifiedClass.getMethod("__writeId", org.bson.types.ObjectId.class), "Should have __writeId method");
        Assert.assertNotNull(modifiedClass.getMethod("__writeCount", long.class), "Should have __writeCount method");
        Assert.assertNotNull(modifiedClass.getMethod("__writeScore", double.class), "Should have __writeScore method");

        Object instance = modifiedClass.getConstructor().newInstance();
        Method writeComputedMethod = modifiedClass.getMethod("__writeComputedValue", String.class);

        try {
            writeComputedMethod.invoke(instance, "test value");
            Assert.fail("Should throw UnsupportedOperationException for read-only property");
        } catch (InvocationTargetException e) {
            Assert.assertTrue(e.getCause() instanceof UnsupportedOperationException,
                    "Should throw UnsupportedOperationException, got: " + e.getCause());
            Assert.assertTrue(
                    e.getCause().getMessage() != null && e.getCause().getMessage().contains("read-only"),
                    "Exception message should mention read-only");
        }
    }

    private String descriptor(Class<?> type, String... typeParameters) {
        String desc = Type.getDescriptor(type);
        if (typeParameters.length > 0) {
            desc = desc.substring(0, desc.length() - 1)
                    + "<" + String.join("", typeParameters) + ">"
                    + ";";
        }
        return desc;
    }

    @SuppressWarnings("unchecked")
    private static <T> TypeData<T> typeDataHelper(Class<T> clazz, TypeData<?>... params) {
        return new TypeData<>(clazz, Arrays.asList(params));
    }
}

File: critter/core/src/test/kotlin/dev/morphia/critter/parser/gizmo/TestGizmoGeneration.kt
package dev.morphia.critter.parser.gizmo

import com.mongodb.client.model.CollationCaseFirst.LOWER
import dev.morphia.annotations.Entity
import dev.morphia.annotations.EntityListeners
import dev.morphia.annotations.Indexes
import dev.morphia.annotations.internal.CollationBuilder.collationBuilder
import dev.morphia.annotations.internal.EntityBuilder.entityBuilder
import dev.morphia.annotations.internal.EntityListenersBuilder.entityListenersBuilder
import dev.morphia.annotations.internal.FieldBuilder.fieldBuilder
import dev.morphia.annotations.internal.IndexBuilder.indexBuilder
import dev.morphia.annotations.internal.IndexOptionsBuilder.indexOptionsBuilder
import dev.morphia.annotations.internal.IndexesBuilder.indexesBuilder
import dev.morphia.critter.Critter.Companion.critterClassLoader
import dev.morphia.critter.parser.Generators.mapper
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator as generator
import dev.morphia.critter.sources.Example
import dev.morphia.mapping.codec.pojo.EntityModel
import dev.morphia.mapping.codec.pojo.PropertyModel
import dev.morphia.mapping.codec.pojo.TypeData
import dev.morphia.mapping.lifecycle.EntityListenerAdapter
import io.quarkus.gizmo.ClassCreator
import io.quarkus.gizmo.MethodDescriptor
import io.quarkus.gizmo.MethodDescriptor.ofMethod
import java.lang.reflect.Modifier
import kotlin.reflect.KClass
import org.objectweb.asm.Type
import org.objectweb.asm.tree.AnnotationNode
import org.testng.Assert.assertEquals
import org.testng.Assert.assertFalse
import org.testng.Assert.assertNotNull
import org.testng.Assert.assertTrue
import org.testng.Assert.fail
import org.testng.annotations.Test

class TestGizmoGeneration {
    @Test
    fun testMapStringExample() {
        var descString = "Ljava/util/Map<Ljava/lang/String;Ldev/morphia/critter/sources/Example;>;"
        var descriptor =
            descriptor(
                Map::class.java,
                descriptor(String::class.java),
                descriptor(Example::class.java)
            )

        assertEquals(descriptor, descString)
        var typeData = typeData(descString)[0]
        assertEquals(
            typeData,
            Map::class.typeData(String::class.typeData(), Example::class.typeData())
        )
    }

    @Test
    fun testListMapStringExample() {
        val descString =
            "Ljava/util/List<Ljava/util/Map<Ljava/lang/String;Ldev/morphia/critter/sources/Example;>;>;"
        val descriptor =
            descriptor(
                List::class.java,
                descriptor(
                    Map::class.java,
                    descriptor(String::class.java),
                    descriptor(Example::class.java)
                )
            )
        assertEquals(descriptor, descString)

        val typeData = typeData(descString)[0]
        assertEquals(
            typeData,
            List::class.typeData(
                Map::class.typeData(String::class.typeData(), Example::class.typeData())
            )
        )
    }

    @Test
    fun testMapOfList() {
        val descString =
            "Ljava/util/Map<Ljava/lang/String;Ljava/util/List<Ldev/morphia/critter/sources/Example;>;>;"
        val descriptor =
            descriptor(
                Map::class.java,
                descriptor(String::class.java),
                descriptor(List::class.java, descriptor(Example::class.java))
            )
        assertEquals(descriptor, descString)
        val typeData = typeData(descriptor)[0]
        assertEquals(
            typeData,
            Map::class.typeData(
                String::class.typeData(),
                List::class.typeData(Example::class.typeData())
            )
        )
    }

    @Test
    fun testPrimitiveArray() {
        val typeData = typeData("[I")[0]
        assertTrue(typeData.array)
    }

    private fun descriptor(type: Class<*>, vararg typeParameters: String): String {
        var desc = Type.getDescriptor(type)
        if (typeParameters.isNotEmpty()) {
            desc =
                desc.dropLast(1) +
                    typeParameters.joinToString("", prefix = "<", postfix = ">") +
                    ";"
        }

        return desc
    }

    private fun KClass<*>.typeData(vararg typeParameters: TypeData<*>): TypeData<*> {
        return TypeData(this.java, listOf(*typeParameters))
    }

    @Test
    fun testAnnotationBuilding() {
        val index = AnnotationNode("Ldev/morphia/annotations/Index;")
        val field = AnnotationNode("Ldev/morphia/annotations/Field;")
        //        field.values = listOf("value", "name")
        index.values = listOf("fields", listOf(field))
        ClassCreator.builder()
            .className("critter.AnnotationTest")
            .superClass(EntityModel::class.java)
            .classOutput { name, data -> critterClassLoader.register(name.replace('/', '.'), data) }
            .build()
            .use {
                val creator = it.getMethodCreator("test", Void::class.java)
                val annotationMethod =
                    ofMethod(
                        EntityModel::class.java.name,
                        "annotation",
                        EntityModel::class.java.name,
                        Annotation::class.java
                    )

                creator.invokeVirtualMethod(
                    annotationMethod,
                    creator.`this`,
                    index.annotationBuilder(creator)
                )
            }
    }

    @Test
    fun testGizmo() {
        generator.generate(Example::class.java)
        critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.AgeModel")
        val nameModel =
            critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.NameModel")
        invokeAll(PropertyModel::class.java, nameModel)
        critterClassLoader.loadClass("dev.morphia.critter.sources.__morphia.example.SalaryModel")
        critterClassLoader
            .loadClass("dev.morphia.critter.sources.__morphia.example.AgeAccessor")
            .getConstructor()
            .newInstance()
        critterClassLoader
            .loadClass("dev.morphia.critter.sources.__morphia.example.NameAccessor")
            .getConstructor()
            .newInstance()
        critterClassLoader
            .loadClass("dev.morphia.critter.sources.__morphia.example.SalaryAccessor")
            .getConstructor()
            .newInstance()
        val loadClass =
            critterClassLoader.loadClass(
                "dev.morphia.critter.sources.__morphia.example.ExampleEntityModel"
            )
        val constructors = loadClass.constructors
        val model: EntityModel = constructors[0].newInstance(mapper) as EntityModel
        validate(model)
    }

    private fun validate(model: EntityModel) {
        val annotation = model.getAnnotation(EntityListeners::class.java)
        assertEquals(
            annotation,
            entityListenersBuilder().value(EntityListenerAdapter::class.java).build()
        )

        assertEquals(
            model.getAnnotation(Entity::class.java),
            entityBuilder().value("examples").build()
        )

        assertEquals(
            model.getAnnotation<Indexes>(Indexes::class.java),
            indexesBuilder()
                .value(
                    indexBuilder()
                        .fields(fieldBuilder().value("name").weight(42).build())
                        .options(
                            indexOptionsBuilder()
                                .partialFilter("partial filter")
                                .collation(collationBuilder().caseFirst(LOWER).build())
                                .build()
                        )
                        .build()
                )
                .build()
        )

        assertEquals(model.collectionName(), "examples")
        assertEquals(model.discriminator(), "Example")
        assertEquals(model.discriminatorKey(), "_t")
        assertEquals(model.type.name, Example::class.java.name)
        assertFalse(model.properties.isEmpty(), "Should have properties")
        assertNotNull(model.idProperty, "Should have an ID property")
        assertFalse(model.isAbstract(), "Should not be abstract")
        assertFalse(model.isInterface(), "Should not be an interface")
        assertTrue(model.useDiscriminator(), "Should use the discriminator")
        assertTrue(model.classHierarchy().isEmpty(), "Should not have a class hierarchy")
    }

    private fun invokeAll(type: Class<*>, klass: Class<*>) {
        val instance = klass.constructors[0].newInstance(null)
        val results =
            type.declaredMethods
                .filter {
                    Modifier.isPublic(it.modifiers) &&
                        !Modifier.isFinal(it.modifiers) &&
                        it.parameterCount == 0
                }
                .filter { it.name !in listOf("hashCode", "toString") }
                .sortedBy { it.name }
                .map { method ->
                    try {
                        klass.getDeclaredMethod(method.name, *method.parameterTypes)
                        null
                    } catch (e: Exception) {
                        e.message
                    }
                }
                .filterNotNull()

        if (results.isNotEmpty()) {
            fail("Missing methods from ${type.name}: \n${results.joinToString("\n")}")
        }
    }

    @Test
    fun testConstructors() {
        val className = "dev.morphia.critter.GizmoSubclass"
        val constructorCall =
            ClassCreator.builder()
                .classOutput { name, data ->
                    critterClassLoader.register(name.replace('/', '.'), data)
                }
                .className("dev.morphia.critter.ConstructorCall")
                .build()
        val fieldCreator =
            constructorCall
                .getFieldCreator("name", String::class.java)
                .setModifiers(Modifier.PUBLIC)
        val constructorCreator = constructorCall.getConstructorCreator(String::class.java)
        constructorCreator.invokeSpecialMethod(
            MethodDescriptor.ofConstructor(Object::class.java),
            constructorCreator.`this`
        )
        constructorCreator.setParameterNames(arrayOf("name"))
        constructorCreator.writeInstanceField(
            fieldCreator.fieldDescriptor,
            constructorCreator.`this`,
            constructorCreator.getMethodParam(0)
        )

        constructorCreator.returnVoid()
        constructorCall.close()
        val newInstance =
            critterClassLoader
                .loadClass("dev.morphia.critter.ConstructorCall")
                .getConstructor(String::class.java)
                .newInstance("here i am")

        val creator =
            ClassCreator.builder()
                .classOutput { name, data ->
                    critterClassLoader.register(name.replace('/', '.'), data)
                }
                .className(className)
                .superClass("dev.morphia.critter.ConstructorCall")
                .build()
        val constructor = creator.getConstructorCreator(String::class.java)
        constructor.invokeSpecialMethod(
            MethodDescriptor.ofConstructor(
                "dev.morphia.critter.ConstructorCall",
                String::class.java
            ),
            constructor.getThis(),
            constructor.getMethodParam(0)
        )
        constructor.setParameterNames(arrayOf("subName"))
        constructor.returnVoid()
        constructor.close()
        creator.close()
        val instance =
            critterClassLoader
                .loadClass(className)
                .getConstructor(String::class.java)
                .newInstance("This is my name")

        assertNotNull(instance)
    }
}
